src/main/java/zd/zero/waifu/motivator/plugin/listeners/WaifuUnitTesterImpl.java

```diff
@@ -1,54 +1,66 @@
 
 public class WaifuUnitTesterImpl implements WaifuUnitTester {
 
     private final MessageBusConnection busConnection;
 
     private final WaifuUnitTester.Listener listener;
 
-    private final Debouncer debouncer;
-
     public WaifuUnitTesterImpl( MessageBusConnection busConnection,
-                                Listener listener,
-                                Debouncer debouncer ) {
+                                Listener listener ) {
         this.busConnection = busConnection;
         this.listener = listener;
-        this.debouncer = debouncer;
     }
 
     @Override
     public void init() {
-        busConnection.subscribe( Notifications.TOPIC, new Notifications() {
+        busConnection.subscribe( SMTRunnerEventsListener.TEST_STATUS, new SMTRunnerEventsAdapter() {
             @Override
-            public void notify( @NotNull Notification notification ) {
-                invokeListener( notification, TestsUIUtil.NOTIFICATION_GROUP.getDisplayId() );
-            }
-        } );
-    }
-
-    void invokeListener( Notification notification, final String NOTIFICATION_GROUP_DISPLAY_ID ) {
-        if ( notification.getGroupId().equals( NOTIFICATION_GROUP_DISPLAY_ID ) ) {
-            debouncer.debounce( () -> {
-                if ( notification.getType() == NotificationType.ERROR ) {
-                    listener.onUnitTestFailed();
-                } else {
+            public void onTestingFinished( @NotNull SMTestProxy.SMRootTestProxy testsRoot ) {
+                if ( testsRoot.wasTerminated() || testsRoot.isInterrupted()
+                    || isCancelledOnExternalProcess( testsRoot ) ) return;
+                if ( testsRoot.isPassed() ) {
                     listener.onUnitTestPassed();
+                } else {
+                    listener.onUnitTestFailed();
                 }
-            } );
-        }
+            }
+        } );
     }
 
     @Override
     public void stop() {
         this.busConnection.disconnect();
     }
 
+    /**
+     * A compatibility hack method to perform checking
+     * for external processes (Gradle test execution)
+     * to check if user has cancelled the test.
+     * <p>
+     * For 202, it should be replaced with
+     * ExternalSystemProcessHandler external = ( ExternalSystemProcessHandler ) testsRoot.getHandler();
+     * ExternalSystemTaskState state = external.getTask().getState();
+     *
+     * @param testsRoot test execution root node
+     * @return true - if it finds 'Build Cancelled'
+     */
+    private boolean isCancelledOnExternalProcess( @NotNull SMTestProxy.SMRootTestProxy testsRoot ) {
+        if ( !( testsRoot.getHandler() instanceof ExternalSystemProcessHandler ) ) return false;
+
+        WaifuTestPrinterAdapter printer = new WaifuTestPrinterAdapter();
+        testsRoot.printOn( printer );
+        testsRoot.flush();
+
+        return printer.getPrintableContent().toLowerCase()
+            .contains( "build cancelled" );
```

Comment:
I know it's bad, please don't punish me.

Reply:
![I ain't even mad](https://i.kym-cdn.com/photos/images/newsfeed/000/369/377/5bd.jpg)
I'm more disappointed about you not choosing to write Kotlin.
😂

Reply:
I'm still married to Java, I'm thinking if I can cheat with Kotlin 😂. Hope Java won't get mad at me.

Reply:
![Kotlin > Java](https://i.imgflip.com/4b5ozs.jpg)

😂